### PR TITLE
Components: update snackbar to use framer motion instead of react spring

### PR DESCRIPTION
**Pull request by gwwar**
This PR experiments with using framer motion instead of react spring for the snackbar component. Since Framer Motion is ~30kb gzipped I also ~externalized it~ exported this from `@wordpress/components` once to avoid bloating each package that it's used in.

One benefit here is that framer motion supports `height: auto` which allows us to remove ref and component state management. 🤔 ~There's quite a bundle size jump here to consider for a single component, so I may explore trying to whittle down.~

Results overall, are are encouraging. I'd like to see if we can update the Persistent List View to use drag and add animations for the drawers opening and closing next.

| Size Impacts |
|----|
| Screen Shot 2021-08-04 at 9 54 40 AM |

#### After - Motion:

https://user-images.githubusercontent.com/1270189/127399049-aa5a91fe-88e9-4783-857e-7531e9adcf8b.mp4

#### After - Reduce Motion

https://user-images.githubusercontent.com/1270189/127397712-3a733583-5e04-485c-bf2f-791e01950e4d.mp4

#### Before - Motion

https://user-images.githubusercontent.com/1270189/127397745-b6b3c29f-5952-4e1a-87c4-cc27564ad7f7.mp4

#### Before - Reduce Motion

https://user-images.githubusercontent.com/1270189/127397786-38325ff9-563c-4446-a2a7-a0b528d19b80.mp4
 

### Testing Instructions

In console type:
```
wp.data.dispatch('core/notices').createNotice(
		'info',
		'Post published.',
		{
			type: 'snackbar',
			actions: [
				{
					onClick: () => {},
					label: 'View post'
				}
			]
		}
	);
```
Or perform an action like publishing or updating a page.

Verify that folks are happy with how framer motion is being exported.

| Verify that there are no regressions in the preview panel |
|-----|
| Screen Shot 2021-08-03 at 12 10 13 PM |

### CODE REVIEW

**Comment by gwwar** [comment]
Leaving some notes for myself:
* play with a slower animation timing for exit
* perhaps remove the x translation when clicking on an action
* see if the bundle size can be shaved down, might need to use m instead (simpler api) https://www.framer.com/docs/guide-reduce-bundle-size/

File: packages/components/src/snackbar/list.js
Comment (youknowriad) [inline packages/components/src/snackbar/list.js:21]: I like framer motion's API very much. This variations one is delightful 👏

File: packages/components/src/snackbar/list.js
Comment (youknowriad) [inline packages/components/src/snackbar/list.js:41]: This is still an animation, I think in "reduce motion", we disable animations entirely in all Gutenberg animations, so I feel we should just do that here as well.

Comment (gwwar) [reply packages/components/src/snackbar/list.js:41]: Good catch 👍

**Comment by gwwar** [comment]
I think I have about equivalent animations now in the branch. The one blocker here is the bump in bundle size by 10% in the components bundle.

So reading through https://www.framer.com/docs/guide-reduce-bundle-size/ it feels like this optimization should be done at the app level. 🤔 I wonder if the size hit is okay here or we need to hold off swapping in framer motion until more components use pieces of it. It's a bit wasteful otherwise in the component package.

File: packages/components/src/snackbar/list.js
Comment (youknowriad) [inline packages/components/src/snackbar/list.js]: After reading the doc, it seems swapping `motion` here with `m` improves the bundle size, did you try it?

Comment (youknowriad) [reply packages/components/src/snackbar/list.js]: Oh I guess, it would force us to add the `LazyMotion` provider, which is not great for packages like you said.

**Review by youknowriad — APPROVED** [review]
LGTM

@gziolo @mtias any thoughts on the bundle size impact.

**Comment by gziolo** [comment]
It's worth mentioning that we use both libraries in other packages and they are bundled with them:

https://github.com/WordPress/gutenberg/blob/85cfab17de6a23a73a6163472d1c9d85581d86c2/packages/edit-post/package.json#L56

By the way, there is a newer version of `framer-motion`:

Screen Shot 2021-07-29 at 11 30 34


https://github.com/WordPress/gutenberg/blob/85cfab17de6a23a73a6163472d1c9d85581d86c2/packages/block-editor/package.json#L68

There is even a new major version of `react-spring`:

Screen Shot 2021-07-29 at 11 31 43

I will leave the user experience aside as I don't see much difference between recorded screencasts.

We should pick one of those libraries and use it consistently. If we pick `framer-motion` we should explore the optimization mentioned:
> see if the bundle size can be shaved down, might need to use m instead (simpler api) https://www.framer.com/docs/guide-reduce-bundle-size/

In their docs they advocate for:

> However, by using the `m` and `LazyMotion` components, we can bring this down significantly, to just under 5kb for the initial render.

If that doesn't work we should consider exposing used APIs from `@wordpress/components` so it isn't bundled in several modules. We do a similar trick with Reakit, example:

https://github.com/WordPress/gutenberg/blob/trunk/packages/components/src/composite/index.js

It isn't great as it gets exposed as public API but we mark them as unstable until we find a better solution:

https://github.com/WordPress/gutenberg/blob/e394488f194472d063a8d6b4dd03f103e7b17919/packages/components/src/index.js#L43-L48

**Comment by gwwar** [comment]
Thanks for the reviews!

> There is even a new major version of react-spring:

The react-spring update PR is here, though I think I'll be able to close that in favor of framer motion.

https://github.com/WordPress/gutenberg/pull/30979

Let's hold on this PR and let me queue up PRs for the two other usages of react-spring. We can then see the total bundle size impacts (and land the PRs within a single Gutenberg release). 

Overall, I'd prefer us using framer-motion for all animations and removing react spring. Usage wise, I'd like us to be able to use the full API over the slimmed down `m` component since that's half of why we like this library, but we may need to see how other packages are affected.

**Comment by mtias** [comment]
Also for reference: #33132.

**Comment by gwwar** [comment]
Here's a draft for one the follow ups for Tree Grid: https://github.com/WordPress/gutenberg/pull/33765

Looks like a very nice update for higher-level abstractions here (though it also makes the PR larger in scope than I originally intended).

**Comment by mtias** [comment]
Looking promising!

File: packages/components/src/snackbar/list.js
Comment (jasmussen) [reply packages/components/src/snackbar/list.js:21]: Same, this is very readable and easy to parse!

**Review by jasmussen — APPROVED** [review]
This is testing well for me, and the code is far more readable. While the framerate of this GIF doesn't show it, it all feels smoother to me as well. I'd love to see framer replace react-spring across the board:

snackbars?

**Comment by gwwar** [comment]
To unblock this one, let me see if I can pull out framer as a chunk that's only loaded once across packages. What I'm seeing from other usages is about a 10% bump in size for individual packages, since it's 20kb more than react-spring.

**Comment by gwwar** [comment]
I have a WIP in b20bceb to externalize the motion framer packager, don't merge in the meantime :D

**Comment by gwwar** [comment]
Ready for another round of reviews 👍 

Much appreciated if folks more familiar with our dependency management can double check to see if I externalized this one properly. cc @gziolo @youknowriad (or anyone else who's done this in the past).

Note that the ~30kb is shuffled to an inline vendor script:
| framer externalized |  externalize framer + drop spring dep from components |
|-----|-----|
| Screen Shot 2021-08-03 at 12 04 54 PM | Screen Shot 2021-08-03 at 12 03 15 PM |

**Comment by youknowriad** [comment]
@gwwar While this works, this make framer motion a public API in WordPress meaning any future upgrades are going to be painful, Can we instead do something similar to what @gziolo shared above reakit https://github.com/WordPress/gutenberg/pull/33717#issuecomment-888969527 and expose maybe an `__unstableMotion` object and API in the components package instead. The advantage of this, is that we'd bundle it only once (components package) but also avoid a public API.

**Comment by gziolo** [comment]
> While this works, this make framer motion a public API in WordPress meaning any future upgrades are going to be painful, Can we instead do something similar to what @gziolo shared above reakit #33717 (comment) and expose maybe an __unstableMotion object and API in the components package instead. The advantage of this, is that we'd bundle it only once (components package) but also avoid a public API.

Right, it's always a major concern that the public API of the 3rd party library changes at some point or we decide that an alternative library would be a better fit. I took a more low-level approach with Reakit, but maybe it'd be better to expose APIs we use under one unstable object. Whatever we decide here, we should adjust the code for Reakit later.

---

Aside: I have just realized that "Compressed Size" GitHub Action should also list vendor files so we could see the full picture 😄 That's why we see only code removed in the comment from the bot: https://github.com/WordPress/gutenberg/pull/33717#issuecomment-887871601. Not sure if it's something that we could fix though as most of the vendor files come from WordPress core. A related config line from the GitHub action:
https://github.com/WordPress/gutenberg/blob/7b9822361eec3715dfbced71714112ddf71b9cc2/.github/workflows/bundle-size.yml#L56

**Comment by gwwar** [comment]
Oh right, not wanting to expose this to Public API 👍 Good point. I'll probably want to add an eslint rule as well to avoid having folks accidentally import a large package in other areas too.

In terms of usage a different unstable package might feel better for users. Though in terms of an actual package it makes little sense.

```jsx
import { motion, AnimatePresence } from '@wordpress/unstableAnimation';
```

It is reasonable to lump this together with the component package for now, since presumably folks need components before they can animate it. Two options for this:

1. import whatever we're using from motion and prefix with __unstable. Works, but there's a chance for namespace collisions to occur. The API for motion is pretty small though, so I don't think this will be an issue.
```jsx
import { __unstableMotion as motion, __unstableAnimatePresence as AnimatePresence } from '@wordpress/components`
```
2. Re-export under an unstable object. Still feels a bit awkward, but things are namespaced. Will need to test any size impacts though.
```jsx
import { __unstableAnimation as Animation } from '@wordpress/components';
const { motion, AnimatePresence } = Animation;
```

**Comment by gwwar** [comment]
I went ahead with option 1: `import { __unstableMotion as motion, __unstableAnimatePresence as AnimatePresence } from '@wordpress/components`. Let me know if folks prefer any name changes here.

| Size Impacts |
|----|
| Screen Shot 2021-08-04 at 9 54 40 AM |

Added ESLint error to avoid folks from importing framer motion directly in other packages:
Screen Shot 2021-08-04 at 9 49 08 AM

**Comment by youknowriad** [comment]
> In terms of usage a different unstable package might feel better for users. Though in terms of an actual package it makes little sense.

Interesting idea, I don't we have explored anything like that yet (publishing an unstable package which I guess means an unstable WP script as well).

I don't have a preference between option 1 and 2. This PR is looking good to me.

**Comment by gwwar** [comment]
Thanks for taking another look @youknowriad! I'll land this tomorrow if folks didn't have any follow up feedback.

File: packages/components/src/index.js
Comment (gziolo) [inline packages/components/src/index.js]: Both methods aren't used outside of `@wordpress/components` so in practice we don't need to expose them for now. It won't have an impact on the bundle size though since they are in use internally.

**Review by gziolo — APPROVED** [review]
Nice, :shipit:

File: packages/components/src/index.js
Comment (gwwar) [reply packages/components/src/index.js]: Sounds good, updated in https://github.com/WordPress/gutenberg/pull/33717/commits/63520aa5dc45f9b774aad47ece74f318d6ac5c35

**Comment by gwwar** [comment]
Thanks for the reviews @youknowriad @gziolo @jasmussen ! ✨

**Pull request merged**
